### PR TITLE
feat(Card): enable one-off styling through BoxProps while maintaining theming

### PR DESCRIPTION
**Pull request by nathanyoung**
# Github Issue or Trello Card
This PR addresses this issue: https://trello.com/c/AeUi6Yvb/1209-card-if-a-box-prop-is-used-do-not-apply-the-card-classes

## Current State

With the recent implementation of component design tokens to enable theming, currently there is no way to allow for single-instance (or one-off) styling of a Card and it's child components. This is due to the fact that the classes applied by the Card.module.scss will always have a higher level of specificity than the utility classes that get applied via BoxProps.

## Solution

This pr makes it so that using any BoxProp whose css attribute matches the Card's will take precendent over the component design token's css attribute. This allows users of the component to use the default styles, use customize the appearance of their cards using the component design tokens, or have one-off styled versions of the Card using BoxProps.

Here's a screenshot of the documentation that illustrates this:

Screen Shot 2022-03-22 at 3 41 23 PM

```jsx
// default appearance
<Card>
  <Card.Section>Default appearance</Card.Section>
  <Card.Section>Default appearance</Card.Section>
  <Card.Footer>footer</Card.Footer>
</Card>

// customized card theme
<Box
  childGap="xl"
  style={{
    '--card-background-color': 'var(--color-brand-tertiary-50)',
    '--card-footer-background-color': 'var(--color-brand-tertiary-900)',
    '--card-border-radius': 'var(--size-border-radius-lg)',
    '--card-section-border-color': 'var(--color-brand-tertiary-100)',
    '--card-box-shadow': '0',
  }}
>
// themed card instance
  <Card>
    <Card.Section>Themed card using component design tokens</Card.Section>
    <Card.Section>Themed card using component design tokens</Card.Section>
    <Card.Footer color="white">footer</Card.Footer>
  </Card>
// one-off styled card within a themed container
  <Card shadow="xl" radius="sm" background="secondary-50">
    <Card.Section>BoxProps are used instead of tokens</Card.Section>
    <Card.Section borderWidth="sm 0 0 0" borderColor="secondary-100">
      BoxProps are used instead of tokens
    </Card.Section>
    <Card.Footer background="secondary-900" color="grey-100">
      footer
    </Card.Footer>
  </Card>
</Box>
```

# What type of change is this?
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Updating documentation.
- [ ] Updating deployment/build pipeline.
- [ ] Updating Project Dependencies
- [ ] Improving or adding to Test Coverage

# Completeness Checklist

- [ ] TESTS: My changes maintain the baseline required test coverage, as specified by code climate analysis.
- [ ] DOCS: All new component work is covered in that component's `Overview` docs.
- [ ] DOCS: All new component work is covered in a component `Playground`.
- [ ] DOCS: All new visual changes or options are covered under relevant components' `VisualTests`.
- [ ] My code has no linting or typescript compile warnings.
- [ ] My work is tied to a Github issue and satisfies the acceptance criteria (if applicable) of the corresponding issue.

# Quality Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# UI Checklist
- [ ] I have conducted visual UAT on my changes/features.
- [ ] My solution works well on desktop, tablet, and mobile browsers.

### CODE REVIEW

File: src/components/Card/Card.module.scss [hunk 1]
```diff
@@ -1,6 +1,6 @@
 .card {
   background-color: var(--card-background-color, var(--color-brand-white-500));
-  border-radius: var(--modal-border-radius, var(--size-border-radius-md));
+  border-radius: var(--card-border-radius, var(--size-border-radius-md));
```
Comment (nathanyoung) [inline src/components/Card/Card.module.scss:3]: bug!

File: src/components/Card/Card.Overview.stories.mdx [hunk 2]
```diff
@@ -332,6 +368,7 @@ Here are some common use cases of `<Card />`.
         '--card-background-color': '--color-brand-white-500',
         '--card-border-radius': '--size-border-radius-md',
         '--card-box-shadow': '--size-box-shadow-sm',
+        '--card-subdued-background-color': '--color-brand-grey-50',
```
Comment (nathanyoung) [inline src/components/Card/Card.Overview.stories.mdx:371]: adding to documentation. was left out before

File: src/components/Card/Card.tsx [hunk 3]
```diff
@@ -13,38 +21,61 @@ export interface CardProps extends BoxProps {
...
+      overflow = 'hidden',
+      display = 'block',
+      radius = undefined,
+      shadow = undefined,
+      width = '100',
+      ...restProps
+    },
+    ref,
+  ) => {
+    const useTheme = background === undefined && radius === undefined && shadow === undefined;
```
Comment (nathanyoung) [inline src/components/Card/Card.tsx:52]: Background, radius, and shadow are themeable component design tokens, so if any of these are passed as props, don't apply the theme styles.

**Comment by arempe93** [comment]
does passing a `background` prop also remove theme styling for `border-radius`?

I think we can split the classes up to keep those defaults

**Comment by nathanyoung** [comment]
@arempe93 yes, at the moment it would strip out all theming styling. I’ll separate it like you suggest, that’s a better way to handle it

**Comment by nathanyoung** [comment]
@arempe93 @juanfabrega I think this is ready for a another review. I added a code example (it's the same as the in the new documentation) in the PR description

File: src/components/Card/Card.module.scss [hunk 4]
```diff
@@ -1,19 +1,30 @@
...
   }
 }
 
-.card-footer {
+.card-footer-background {
   background-color: var(--card-footer-background-color, var(--color-brand-grey-50));
-  border-color: var(--card-section-border-color, var(--color-brand-grey-100));
 }
 
-.card-section + .card-section {
```
Comment (juanfabrega) [inline src/components/Card/Card.module.scss:17]: Why is this not necessary moving forward?

File: src/components/Card/components/CardFooter/CardFooter.tsx [hunk 5]
```diff
@@ -1,34 +1,60 @@
...
 export interface CardFooterProps extends BoxProps {
   /**
    * Contents of the Footer.
    */
   children?: ReactNode;
+  /**
+   * If defined as a prop, this value will take higher precedence than the corresponding component design token value
+   * Any valid [brand color token](/?path=/story/design-tokens-design-tokens--page#color), or a `url()` for an image
+   */
+  background?: BrandColor;
```
Comment (juanfabrega) [inline src/components/Card/components/CardFooter/CardFooter.tsx:16]: To ensure consistency across these props in the Box and the CardFooter, I would change these to pull directly from BoxProps. E.G:

`background?: BoxProps['background'];`

File: src/components/Card/components/CardFooter/CardFooter.tsx [hunk 6]
```diff
@@ -1,34 +1,60 @@
...
+  /**
+   * If defined as a prop, this value will take higher precedence than the corresponding component design token value
+   * Any valid [brand color token](/?path=/story/design-tokens-design-tokens--page#color) for the border color
+   * Or a responsive prop with BrandColor for each breakpoint.
+   */
+  borderColor?: BrandColor;
```
Comment (juanfabrega) [inline src/components/Card/components/CardFooter/CardFooter.tsx:22]: Same comment as background.

**Review by juanfabrega — APPROVED** [review]
I had some minor comments but no showstoppers. 

I think a topic for later discussion is how this fits into the library moving forward, since it seems to complicate the class system for components. I can't think of other clear solutions at moment but doing this for larger components would create quite a bit of friction.

File: src/components/Card/Card.module.scss (hunk 4 above)
Comment (nathanyoung) [reply src/components/Card/Card.module.scss:17]: it is still necessary, but written this way

```
.card-section-border {
  + .card-section-border {
    border-top: 1px solid var(--card-section-border-color, var(--color-brand-grey-100));
  }
}
```

**Review by arempe93 — APPROVED** [review]

**Pull request merged**